src/sage/schemes/hyperelliptic_curves/jacobian_endomorphisms.py

```diff
@@ -0,0 +1,465 @@
+is the case, one says that `J` is ``generic``.
+
+Both of these are important attributes of the Jacobian J.
+
+The algorithms of these two methods are implementations of Algorithms 4.10 and
+4.15 from Lombardo's paper [Lom2019]_.
+
+The following examples have been verified with the corresponding LMFDB entries.
+
+EXAMPLES::
+
+Here is an example of a generic Jacobian; the LMFDB label of the curve is
+249.a.249.1::
+
+    sage: R.<x> = QQ[]
+    sage: f = x^6 + 2*x^3 + 4*x^2 + 4*x + 1
+    sage: C = HyperellipticCurve(f)
+    sage: A = C.jacobian()
+    sage: E = A.endomorphism_ring()
+    sage: E.is_absolutely_trivial()
+    True
+
+Here is an example of a Jacobian whose endomorphism algebra is a field but not
+the rational number field; the LMFDB label of the curve is 529.a.529.1::
+
+    sage: f = x^6 - 4*x^5 + 2*x^4 + 2*x^3 + x^2 + 2*x + 1
+    sage: C = HyperellipticCurve(f)
+    sage: A = C.jacobian()
+    sage: E = A.endomorphism_ring()
+    sage: E.is_absolutely_trivial()
+    False
+
+Here is an example of a Jacobian whose endomorphism algebra is not a field;
+the LMFDB label of the curve is 169.a.169.1::
+
+    sage: f = x^6 + 4*x^5 + 6*x^4 + 2*x^3 + x^2 + 2*x + 1
+    sage: C = HyperellipticCurve(f)
+    sage: A = C.jacobian()
+    sage: E = A.endomorphism_ring()
+    sage: E.is_absolutely_trivial()
+    False
+
+.. WARNING::
+
+    There is a very small chance that the algorithms return ``False`` for the
+    two methods described above when in fact one or both of them are ``True``.
+    In this case, as explained in the discussion immediately preceding
+    Algorithm 4.15 of [Lom2019]_, this can be established by increasing the
+    optional `B` parameter of the two methods. Mathematically, the algorithms
+    given the correct answer only in the limit as `B \to \infty`, although in
+    practice `B = 200` was sufficient to correctly verify every single entry
+    in the LMFDB.
+
+AUTHORS:
+
+- Barinder S. Banwait and Davide Lombardo (2021-06-09): initial version
+
+"""
+
+# ****************************************************************************
+#       Copyright (C) 2021 Barinder S. Banwait and Davide Lombardo
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#                  https://www.gnu.org/licenses/
+# ****************************************************************************
+
+from sage.structure.sage_object import SageObject
+from sage.rings.all import QQ, ZZ, PolynomialRing, FiniteField, NumberField
+from sage.misc.lazy_import import lazy_import
+from sage.rings.fast_arith import prime_range
+from sage.arith.all import gcd
+lazy_import('sage.interfaces.genus2reduction', ['genus2reduction', 'Genus2reduction'])
+
+
+def satisfies_coefficient_condition(g,p):
```

Comment:
white space after `g,`.
See: https://www.python.org/dev/peps/pep-0008/#whitespace-in-expressions-and-statements

Reply:
Done